### PR TITLE
CVSL-1524 Adding creation of hardstop licence

### PR DESCRIPTION
**Pull request by andrewrlee**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceCreationService.kt [hunk 1]
```diff
@@ -77,37 +76,92 @@ class LicenceCreationService(
...
+    val responsibleCom = staffRepository.findByStaffIdentifier(currentResponsibleOfficerDetails.staffId)
+      ?: createCom(currentResponsibleOfficerDetails.staffId)
+
+    val createdBy = staffRepository.findByUsernameIgnoreCase(username) as PrisonCaseAdministrator?
+      ?: error("Staff with username $username not found")
+
+    val timedOutLicence: CrdLicence? = licenceRepository.findByBookingIdAndStatusCodeOrderByDateCreatedDesc(
+      nomisRecord.bookingId.toLong(),
+      TIMED_OUT,
+    )
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceCreationService.kt:114]: Hooking up a reference to any previous TIMED_OUT licence

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceCreationService.kt [hunk 2]
```diff
@@ -77,37 +76,92 @@ class LicenceCreationService(
...
 
-    val licenceEntity = licenceRepository.saveAndFlush(licence)
-    val createLicenceResponse = transformToLicenceSummary(licenceEntity)
+    val createdLicence = licenceRepository.saveAndFlush(licence)
 
-    val standardConditions = licencePolicyService.getStandardConditionsForLicence(licenceEntity)
+    val standardConditions = licencePolicyService.getStandardConditionsForLicence(createdLicence)
     standardConditionRepository.saveAllAndFlush(standardConditions)
 
+    // Need to add the additional condition here
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceCreationService.kt:134]: I'll add this as a follow up

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/util/LicenceKind.kt [hunk 3]
```diff
@@ -4,25 +4,29 @@ import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.util.LicenceStatus.
 import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.util.LicenceStatus.VARIATION_IN_PROGRESS
 
 enum class LicenceKind(
-  val initialStatus: LicenceStatus,
-  val copyEventType: LicenceEventType,
-  val submittedEventType: LicenceEventType,
+  val initialStatus: () -> LicenceStatus,
+  val creationEventType: () -> LicenceEventType,
+  val copyEventType: () -> LicenceEventType,
+  val submittedEventType: () -> LicenceEventType,
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/util/LicenceKind.kt:10]: These are now lazy to allow state checking in the unlikely case of an unsupported transition

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/privateApi/LicenceController.kt [hunk 4]
```diff
@@ -93,8 +95,9 @@ class LicenceController(
   fun createLicence(
     @RequestBody @Valid
     request: CreateLicenceRequest,
-  ): LicenceSummary {
-    return licenceCreationService.createLicence(request.nomsId)
+  ): LicenceSummary = when (request.type) {
+    CRD -> licenceCreationService.createLicence(request.nomsId)
+    HARD_STOP -> licenceCreationService.createHardStopLicence(request.nomsId)
   }
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/privateApi/LicenceController.kt:101]: used existing endpoint and the new type field determines which type of licence to create

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/model/request/CreateLicenceRequest.kt [hunk 5]
```diff
@@ -3,9 +3,16 @@ package uk.gov.justice.digital.hmpps.createandvaryalicenceapi.model.request
...
+}
+
 @Schema(description = "Request object for creating a new licence")
 data class CreateLicenceRequest(
   @Schema(description = "The prison nomis identifier for this offender", example = "A1234AA")
   @NotNull
   val nomsId: String,
+
+  @Schema(description = "The type of licence to create", example = "CRD")
+  val type: LicenceType = LicenceType.CRD,
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/model/request/CreateLicenceRequest.kt:17]: default to CRD to allow existing frontend to work as is

**Review by nishanthmahasamudram — APPROVED** [review]
LGTM

**Pull request merged**